# Regression, VIF, Correlation
#### data: DISCRIM from wooldridge

### Regression Model
##### log(psoda) = β0 + β1prpblck + β2log(income) + β3prppov + u
Test if β1 is statistically different from zero at the 5% level against a two-sided alternative

In [9]:
#import packages
#import warnings
#warnings.simplefilter(action='ignore', category='FutureWarning')
#warnings.filterwarnings('ignore')

import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats
import pandas as pd
import seaborn as sns
import statistics
import statsmodels.formula.api as smf

#import data
import wooldridge as woo 
data = woo.data('DISCRIM') 
data.head()

psoda  pfries  pentree  wagest  nmgrs  nregs  hrsopen   emp  psoda2  \
0   1.12    1.06     1.02    4.25    3.0    5.0     16.0  27.5    1.11   
1   1.06    0.91     0.95    4.75    3.0    3.0     16.5  21.5    1.05   
2   1.06    0.91     0.98    4.25    3.0    5.0     18.0  30.0    1.05   
3   1.12    1.02     1.06    5.00    4.0    5.0     16.0  27.5    1.15   
4   1.12     NaN     0.49    5.00    3.0    3.0     16.0   5.0    1.04   

   pfries2  ...  county    lpsoda   lpfries    lhseval    lincome  ldensity  \
0     1.11  ...      18  0.113329  0.058269  11.906993  10.704008  8.301521   
1     0.89  ...      18  0.058269 -0.094311  11.906993  10.704008  8.301521   
2     0.94  ...      12  0.058269 -0.094311  12.038836  10.625319  9.341369   
3     1.05  ...      10  0.113329  0.019803  12.052921  10.827071  9.029418   
4     1.01  ...      10  0.113329       NaN  12.425610  11.188399  6.579251   

   NJ  BK  KFC  RR  
0   1   0    0   1  
1   1   1    0   0  
2   1   1    0   0  
3   1   0    0   1  
4   1   1    0   0  

[5 rows x 37 columns]

In [40]:
results = smf.ols('lpsoda ~ prpblck + lincome + prppov', data).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 lpsoda   R-squared:                       0.087
Model:                            OLS   Adj. R-squared:                  0.080
Method:                 Least Squares   F-statistic:                     12.60
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           6.92e-08
Time:                        10:29:54   Log-Likelihood:                 439.04
No. Observations:                 401   AIC:                            -870.1
Df Residuals:                     397   BIC:                            -854.1
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.4633      0.294     -4.982      0.000      -2.041      -0.886
prpblck        0.0728      0.031      2.373      0.018       0.013       0.133
lincome        0.1370      0.027      5.119      0.000       0.084       0.190
prppov         0.3804      0.133      2.864      0.004       0.119       0.641
==============================================================================
Omnibus:                       12.002   Durbin-Watson:                   1.727
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               24.056
Skew:                          -0.014   Prob(JB):                     5.97e-06
Kurtosis:                       4.200   Cond. No.                         834.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

##### Beta 1 is statistically significant at 5% level, as it has a p-value of 0.018. And its confidence interval does not include 0. 
##### It is not statistically significant at 1% level because 0.018 > 0.01.

### Correlation  between  log(income)  and  prppov, and VIF

In [41]:
# Correlation between log(income) and prppov
df = pd.DataFrame({'X':data['lincome'], 'Y':data['prppov']})
df['X'].corr(df['Y'])

-0.8384669885470059

In [42]:
# VIF
import patsy as pt
import statsmodels.stats.outliers_influence as smo

# Get the design matrix
y, X = pt.dmatrices('lpsoda ~ prpblck + lincome + prppov', data = data,
                   return_type = 'dataframe')

# X.head()
#  Pull the number of regressors (+ intercept)
k = X.shape[1]

# create an empty matrix to store results
VIF = np.empty(k)

# Loop for each regressor (+ intercept)
for i in range(k):
    
    # calculate the VIF for each
    VIF[i] = smo.variance_inflation_factor(X.values, i)

print('VIF:', VIF)

VIF: [5.22495429e+03 1.92215973e+00 3.51872158e+00 4.91522152e+00]


###### As the VIF values of three predictors are less than 10, they are statistically significant. 

###### P-value of log(income) is 0.000, of prppov is 0.004. 

### Add the variable log(hseval)

In [43]:
results2 = smf.ols('lpsoda ~ prpblck + lincome + prppov + lhseval', data).fit()
results2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 lpsoda   R-squared:                       0.184
Model:                            OLS   Adj. R-squared:                  0.176
Method:                 Least Squares   F-statistic:                     22.31
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.24e-16
Time:                        10:37:31   Log-Likelihood:                 461.55
No. Observations:                 401   AIC:                            -913.1
Df Residuals:                     396   BIC:                            -893.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.8415      0.292     -2.878      0.004      -1.416      -0.267
prpblck        0.0976      0.029      3.334      0.001       0.040       0.155
lincome       -0.0530      0.038     -1.412      0.159      -0.127       0.021
prppov         0.0521      0.134      0.388      0.699      -0.212       0.317
lhseval        0.1213      0.018      6.860      0.000       0.087       0.156
==============================================================================
Omnibus:                       16.452   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               40.077
Skew:                          -0.016   Prob(JB):                     1.98e-09
Kurtosis:                       4.548   Cond. No.                     1.31e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.31e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

###### The coefficient of log(hseval) is 0.1213, which indicates that for a 1% increase in the median housing value, the price of medium soda goes up by 0.1213%. 
###### The p-value is 0.000. We can reject the null hypothesis and conclude that the coefficient of log(hseval) is not zero. 

### Joint Significance of Predictors

###### Individually, the p-values of log(income) and prppov became large and they were not statistically significant. 

In [44]:
# Jointly:
# write each of your hypotheses in a list
hypotheses = ['lincome = 0', 'prppov = 0']

# use the f-test method included in sm.ols().fit() objects
results2.f_test(hypotheses) 

<class 'statsmodels.stats.contrast.ContrastResults'>
<F test: F=3.5226849440225143, p=0.030448605108275816, df_denom=396, df_num=2>

###### Thus, jointly, they are statistically significant at a 5% significance level with a small p-value of 0.03. 

###### The second regression model explains more portion of the observations of the dataset. It's R-squared value is 0.176 which is larger than that of the previous model which only explains 0.08 of the data. Therefore the second model is more reliable. 

## data: VOTE1 dataset from the Wooldridge

##### voteA = β0 + β1prtystrA + β2log(expendA) + β3democA + β4log(expendB) + u


### OLS residuals, ˆui, and regress these on all of the independent variables.

In [54]:
vote = woo.data('VOTE1') 
results3 = smf.ols('voteA ~ prtystrA + lexpendA + democA + lexpendB', vote).fit()
results3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.801
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     169.2
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           8.09e-58
Time:                        21:57:11   Log-Likelihood:                -593.20
No. Observations:                 173   AIC:                             1196.
Df Residuals:                     168   BIC:                             1212.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     37.6614      4.736      7.952      0.000      28.312      47.011
prtystrA       0.2519      0.071      3.534      0.001       0.111       0.393
lexpendA       5.7793      0.392     14.750      0.000       5.006       6.553
democA         3.7929      1.407      2.697      0.008       1.016       6.570
lexpendB      -6.2378      0.397    -15.694      0.000      -7.022      -5.453
==============================================================================
Omnibus:                        6.304   Durbin-Watson:                   1.525
Prob(Omnibus):                  0.043   Jarque-Bera (JB):                6.030
Skew:                           0.448   Prob(JB):                       0.0491
Kurtosis:                       3.182   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [55]:
# Obtain OLS residuals
# pull out squared residuals
vote["res2"] = results3.resid**2

# try to predict the squared residuals using a linear combination of our variables
aux_reg = smf.ols('res2 ~ prtystrA + lexpendA + democA + lexpendB', vote).fit()

aux_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   res2   R-squared:                       0.053
Model:                            OLS   Adj. R-squared:                  0.030
Method:                 Least Squares   F-statistic:                     2.330
Date:                Thu, 17 Nov 2022   Prob (F-statistic):             0.0581
Time:                        21:57:18   Log-Likelihood:                -1003.7
No. Observations:                 173   AIC:                             2017.
Df Residuals:                     168   BIC:                             2033.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    113.9635     50.815      2.243      0.026      13.645     214.282
prtystrA      -0.2993      0.765     -0.391      0.696      -1.809       1.211
lexpendA     -10.3057      4.204     -2.451      0.015     -18.605      -2.006
democA        15.6192     15.091      1.035      0.302     -14.174      45.412
lexpendB      -0.0514      4.265     -0.012      0.990      -8.470       8.368
==============================================================================
Omnibus:                      127.443   Durbin-Watson:                   1.913
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1080.188
Skew:                           2.761   Prob(JB):                    2.76e-235
Kurtosis:                      13.925   Cond. No.                         429.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###### Adjusted R-squared is 0.03 because the residuals have little relationship with independent variables. 

### Breusch-Pagan test for heteroskedasticity and the F statistic

In [48]:
# Get the regression f-statistic (f-test version)
f = aux_reg.fvalue
fp = aux_reg.f_pvalue

print("The F-Statistic for the Auxiliary Regression is: "+ str(f) +" and the P-Value is: "+ str(fp))

The F-Statistic for the Auxiliary Regression is: 2.33011282674083 and the P-Value is: 0.058057501107011425


###### Therefore we fail to reject the null hypothesis because we have a p-value of 0.058 which is larger than the 5% level. And there is not heteroscedasticity in this model.

### White test for heteroskedasticity, and the F statistic

In [50]:
import statsmodels.api as sm

y, X = pt.dmatrices('voteA ~ prtystrA + lexpendA + democA + lexpendB', vote,
                   return_type = 'dataframe')

# Order is Lm Test statistic, LM P-value, F-stat, F-Pvalue
sm.stats.diagnostic.het_white(results3.resid, X)

(31.10151403188262,
 0.003258263798817035,
 2.680757572966332,
 0.001973326570119174)

###### The p-value is 0.00197 which is smaller than 5%. Therefore we can conclude that there exists heteroscedasticity in our model.

### Robust standard errors (HC0) and the FGLS procedure

In [67]:
# Model 1: replace log(expendB) with expendB
robust_reg = smf.ols('voteA ~ prtystrA + lexpendA + democA + expendB', vote).fit(cov_type = 'HC0')
robust_reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.730
Model:                            OLS   Adj. R-squared:                  0.723
Method:                 Least Squares   F-statistic:                     78.54
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.92e-37
Time:                        22:21:06   Log-Likelihood:                -619.80
No. Observations:                 173   AIC:                             1250.
Df Residuals:                     168   BIC:                             1265.
Df Model:                           4                                         
Covariance Type:                  HC0                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.4524      4.532      0.762      0.446      -5.431      12.336
prtystrA       0.3815      0.083      4.584      0.000       0.218       0.545
lexpendA       6.7333      0.613     10.985      0.000       5.532       7.935
democA         4.9870      1.720      2.899      0.004       1.616       8.358
expendB       -0.0280      0.003     -9.871      0.000      -0.034      -0.022
==============================================================================
Omnibus:                        2.307   Durbin-Watson:                   1.496
Prob(Omnibus):                  0.316   Jarque-Bera (JB):                1.985
Skew:                           0.256   Prob(JB):                        0.371
Kurtosis:                       3.118   Cond. No.                     2.88e+03
==============================================================================

Notes:
[1] Standard Errors are heteroscedasticity robust (HC0)
[2] The condition number is large, 2.88e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [66]:
# pull out squared residuals
vote["res1"] = robust_reg.resid**2

# try to predict the squared residuals using a linear combination of our variables
aux_reg1 = smf.ols('res1 ~ prtystrA + lexpendA + democA + expendB', vote).fit()

# Get the regression f-statistic (f-test version)
f1 = aux_reg1.fvalue
fp1 = aux_reg1.f_pvalue

print("The F-Statistic for the Auxiliary Regression is: "+ str(f1) +" and the P-Value is: "+ str(fp1))

The F-Statistic for the Auxiliary Regression is: 3.408306759939193 and the P-Value is: 0.010403816019878638


###### Therefore we fail to reject the null hypothesis because we have a p-value of 0.0104 which is smaller than the 5% level. And there is  heteroscedasticity in this model.

This new model explains 72.3% of the data. The statistical significance of coefficients still holds. 

In [70]:
# Model 2: replace log(expendA) with expendA
results4 = smf.ols('voteA ~ prtystrA + expendA + democA + lexpendB', vote).fit()
results4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.730
Model:                            OLS   Adj. R-squared:                  0.723
Method:                 Least Squares   F-statistic:                     113.5
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.09e-46
Time:                        22:30:07   Log-Likelihood:                -619.68
No. Observations:                 173   AIC:                             1249.
Df Residuals:                     168   BIC:                             1265.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     53.1161      5.642      9.414      0.000      41.977      64.255
prtystrA       0.3769      0.081      4.654      0.000       0.217       0.537
expendA        0.0275      0.003     10.764      0.000       0.022       0.033
democA         7.1769      1.588      4.519      0.000       4.042      10.312
lexpendB      -6.8565      0.476    -14.410      0.000      -7.796      -5.917
==============================================================================
Omnibus:                        3.767   Durbin-Watson:                   1.551
Prob(Omnibus):                  0.152   Jarque-Bera (JB):                3.275
Skew:                           0.247   Prob(JB):                        0.194
Kurtosis:                       2.542   Cond. No.                     3.60e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.6e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [69]:
vote["ehatsq"] = results4.resid**2

# estimate weights
w_est = smf.ols('np.log(ehatsq) ~ prtystrA + expendA + democA + lexpendB', data = vote).fit()

vari = np.exp(w_est.fittedvalues) #estimated variances
w = 1/vari**2

fgls =smf.wls('voteA ~ prtystrA + expendA + democA + lexpendB', vote, weights = w).fit()

fgls.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            WLS Regression Results                            
==============================================================================
Dep. Variable:                  voteA   R-squared:                       0.764
Model:                            WLS   Adj. R-squared:                  0.758
Method:                 Least Squares   F-statistic:                     135.7
Date:                Thu, 17 Nov 2022   Prob (F-statistic):           1.54e-51
Time:                        22:26:54   Log-Likelihood:                -633.04
No. Observations:                 173   AIC:                             1276.
Df Residuals:                     168   BIC:                             1292.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     49.4383      5.978      8.270      0.000      37.637      61.240
prtystrA       0.4336      0.085      5.119      0.000       0.266       0.601
expendA        0.0293      0.003     10.879      0.000       0.024       0.035
democA         7.0983      1.589      4.466      0.000       3.961      10.236
lexpendB      -6.8048      0.453    -15.018      0.000      -7.699      -5.910
==============================================================================
Omnibus:                        9.747   Durbin-Watson:                   1.638
Prob(Omnibus):                  0.008   Jarque-Bera (JB):               11.935
Skew:                           0.398   Prob(JB):                      0.00256
Kurtosis:                       4.010   Cond. No.                     3.50e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.5e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""